File: Julia/bert-pooler.jl
tanh_act(x) = tanh.(x)

mutable struct BertPooler
    hidden_size::Int
    W
    b
end

function BertPooler(hidden_size::Int)
    W = randn(hidden_size, hidden_size) .* 0.02
    b = zeros(hidden_size)
    BertPooler(hidden_size, W, b)
end

function forward(pooler::BertPooler, hidden_states)
    cls_token_tensor = hidden_states[:, 1, :]
    pooled_output = cls_token_tensor * pooler.W .+ pooler.b
    tanh_act(pooled_output)
end

mutable struct SequenceClassifier
    pooler::BertPooler
    dropout_prob::Float64
    classifier
    bias
end

function SequenceClassifier(hidden_size::Int, num_classes::Int; dropout_prob::Float64=0.1)
    pooler = BertPooler(hidden_size)
    classifier = randn(hidden_size, num_classes) .* 0.02
    bias = zeros(num_classes)
    SequenceClassifier(pooler, dropout_prob, classifier, bias)
end

function forward(model::SequenceClassifier, hidden_states; training::Bool=true)
    pooled_output = forward(model.pooler, hidden_states)
    if training
        mask = rand(size(pooled_output)) .> model.dropout_prob
        pooled_output = (pooled_output .* mask) ./ (1.0 - model.dropout_prob)
    end
    pooled_output * model.classifier .+ model.bias
end

File: R/bert-pooler.R
tanh_act <- function(x) {
  tanh(x)
}

BertPooler <- setRefClass(
  "BertPooler",
  fields = list(
    hidden_size = "numeric",
    W = "matrix",
    b = "numeric"
  ),
  methods = list(
    initialize = function(hidden_size) {
      hidden_size <<- hidden_size
      W <<- matrix(rnorm(hidden_size * hidden_size, sd = 0.02), nrow = hidden_size, ncol = hidden_size)
      b <<- numeric(hidden_size)
    },
    forward = function(hidden_states) {
      cls_token_tensor <- hidden_states[, 1, ]
      pooled_output <- cls_token_tensor %*% W + b
      tanh_act(pooled_output)
    }
  )
)

SequenceClassifier <- setRefClass(
  "SequenceClassifier",
  fields = list(
    pooler = "BertPooler",
    dropout_prob = "numeric",
    classifier = "matrix",
    bias = "numeric"
  ),
  methods = list(
    initialize = function(hidden_size, num_classes, dropout_prob = 0.1) {
      pooler <<- BertPooler$new(hidden_size)
      dropout_prob <<- dropout_prob
      classifier <<- matrix(rnorm(hidden_size * num_classes, sd = 0.02), nrow = hidden_size, ncol = num_classes)
      bias <<- numeric(num_classes)
    },
    forward = function(hidden_states, training = TRUE) {
      pooled_output <- pooler$forward(hidden_states)
      if (training) {
        mask <- matrix(runif(length(pooled_output)) > dropout_prob, nrow = nrow(pooled_output))
        pooled_output <- (pooled_output * mask) / (1.0 - dropout_prob)
      }
      pooled_output %*% classifier + bias
    }
  )
)
